### PR TITLE
Coupons: Create coupon

### LINKED ISSUE #6492: Coupons: Add coupon creation flow
This flow should reuse the UI from the edit flow:

Screen Shot 2022-05-20 at 12 08 39

### PR DESCRIPTION
**Pull request by ThomazFB**
Summary
==========
Fix issue #6492 by adapting the Coupon list view to be able to create Coupons using the infrastructure already available inside the `AddEditCoupons` view, and also taking advantage of the work done in #6493 to display the Coupon creation success view.

Screenshots
==========
https://user-images.githubusercontent.com/5920403/172998884-b12e6475-7afb-4676-b7f4-17f460cc7748.mp4

How to Test
==========
## Scenario 1
1. Make sure the Coupons experimental toggle is activated
2. Go to the Coupon list, make sure there's no coupon created for your site yet
3. Verify that the Coupon list view now displays information about creating coupons with a button, but also a "plus" icon button in the top right corner of the navigation bar
4. Select any of the two coupon creation buttons, verify the discount type bottom sheet, and select a type
5. Verify that every field of the Coupon creation form is empty as expected and interactable
6. Verify that you can't create a coupon without a coupon code defined
7. Once all is set, hit `Create` and a Coupon success view is displayed with a working share button and a close button
8. Hit close and verify that your new coupon is displayed correctly in the Coupon list

## Scenario 2
1. Make sure the Coupons experimental toggle is activated
2. Go to the Coupon list, now with at least one coupon configured beforehand
3. Verify that the coupon list is populated and expected and the "plus" button in the top right corner is displayed alongside the search button
4. Select the "plus" icon button, verify the discount type bottom sheet, and select a type
5. Go through the entire coupon creation flow and the creation success view
6. Hit close and verify that your new coupon is displayed correctly in the Coupon list

## Scenario 3
1. Make sure the Coupons experimental toggle is activated
2. Go to the Coupon list, now with JUST ONE coupon configured beforehand
3. Select the single coupon in the list and once inside the Coupon details, delete it
4. Now back to the coupon list, verify that the coupon creation empty view is back there to start a fresh coupon creation flow

Update release notes:

- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

### CODE REVIEW

File: WooCommerce/Classes/ViewRelated/Coupons/Coupon+String.swift [hunk 1]
```diff
@@ -0,0 +1,29 @@
+import Foundation
+import Yosemite
+
+extension Coupon {
```
Comment (ThomazFB) [inline WooCommerce/Classes/ViewRelated/Coupons/Coupon+String.swift:4]: This code was previously inside the Coupon details ViewModel, but since we now need this logic in the Coupon list also, I decided to go with code reuse and made the share message something generated from within the Coupon through this extension, not sure if this would be an acceptable way to do that using Swift tho.

File: WooCommerce/Classes/Analytics/WooAnalyticsStat.swift [hunk 2]
```diff
@@ -586,6 +587,9 @@ public enum WooAnalyticsStat: String {
     case couponUpdateInitiated = "coupon_update_initiated"
     case couponUpdateSuccess = "coupon_update_success"
     case couponUpdateFailed = "coupon_update_failed"
+    case couponCreationInitiated = "coupon_creation_initiated"
+    case couponCreationSuccess = "coupon_creation_success"
+    case couponCreationFailed = "coupon_creation_failed"
```
Comment (ThomazFB) [inline WooCommerce/Classes/Analytics/WooAnalyticsStat.swift:592]: For the coupon creation tracks, I followed the very same pattern used in the update part. Let me know if the tracks should be tackled in a different PR.

File: WooCommerce/Classes/ViewRelated/Coupons/Add and Edit Coupons/AddEditCouponViewModel.swift [hunk 3]
```diff
@@ -233,7 +246,43 @@ final class AddEditCouponViewModel: ObservableObject {
...
+    func completeCouponAddEdit(coupon: Coupon, onUpdateFinished: @escaping () -> Void) {
+        if isCreationFlow {
+            createCoupon(coupon: coupon)
+        } else {
+            updateCoupon(coupon: coupon, onUpdateFinished: onUpdateFinished)
+        }
+    }
+
+    private func createCoupon(coupon: Coupon) {
+        ServiceLocator.analytics.track(.couponCreationInitiated)
```
Comment (ThomazFB) [inline WooCommerce/Classes/ViewRelated/Coupons/Add and Edit Coupons/AddEditCouponViewModel.swift]: Since there's nothing getting updated here, but an entire coupon made from scratch instead, I thought that it would make sense to just trigger the couponCreationInitiated by itself, with no additional parameters.

File: WooCommerce/Classes/ViewRelated/Coupons/CouponListViewController.swift [hunk 4]
```diff
@@ -201,6 +239,7 @@ extension CouponListViewController: UITableViewDelegate {
             self.navigationController?.popViewController(animated: true)
             let notice = Notice(title: Localization.couponDeleted, feedbackType: .success)
             self.noticePresenter.enqueue(notice: notice)
+            self.viewModel.onCouponDeleted()
```
Comment (ThomazFB) [inline WooCommerce/Classes/ViewRelated/Coupons/CouponListViewController.swift]: I noticed that when we delete the last coupon from the list when don't notify the view model about it so it can change the list state back to `.empty`, leading to a blank view with no information. Since we now have a new empty list view, I decided to fix this here by defining the `onCouponDeleted` function.

File: WooCommerce/Classes/Analytics/WooAnalyticsStat.swift (hunk 2 above)
Comment (itsmeichigo) [reply WooCommerce/Classes/Analytics/WooAnalyticsStat.swift:592]: I think it's fine to add tracks in this PR 👍 However I see that I missed the event for `couponsListCreateTapped` in my analytics plan p91TBi-7vL-p2 - and I have some slight typos on the event names so I'll update the post to reflect these.

File: WooCommerce/Classes/ViewRelated/Coupons/Add and Edit Coupons/AddEditCouponViewModel.swift [hunk 5]
```diff
@@ -15,6 +15,8 @@ final class AddEditCouponViewModel: ObservableObject {
 
     private let onCompletion: ((Result<Coupon, Error>) -> Void)
 
+    private let isCreationFlow: Bool
```
Comment (itsmeichigo) [inline WooCommerce/Classes/ViewRelated/Coupons/Add and Edit Coupons/AddEditCouponViewModel.swift]: I believe we already have a similar variable `editingOption`, which can be used for checking the state of the view. Maybe this new variable is redundant?

File: WooCommerce/Classes/ViewRelated/Coupons/Add and Edit Coupons/AddEditCouponViewModel.swift [hunk 6]
```diff
@@ -29,6 +31,14 @@ final class AddEditCouponViewModel: ObservableObject {
         }
     }
 
+    var addEditCouponButtonText: String {
```
Comment (itsmeichigo) [inline WooCommerce/Classes/ViewRelated/Coupons/Add and Edit Coupons/AddEditCouponViewModel.swift:36]: Nit: We should probably add a comment to clarify what the value is for?

File: WooCommerce/Classes/ViewRelated/Coupons/Add and Edit Coupons/AddEditCouponViewModel.swift (hunk 3 above)
Comment (itsmeichigo) [reply WooCommerce/Classes/ViewRelated/Coupons/Add and Edit Coupons/AddEditCouponViewModel.swift]: In my tracking plan p91TBi-7vL-p2, I added properties for this event in case we want to check for any specific field in the creation flow. Do you think they're redundant?

File: WooCommerce/Classes/ViewRelated/Coupons/CouponDetails/CouponDetailsViewModel.swift [hunk 7]
```diff
@@ -97,13 +97,7 @@ final class CouponDetailsViewModel: ObservableObject {
...
     ///
     var shareMessage: String {
-        if coupon.productIds.isNotEmpty ||
-            coupon.productCategories.isNotEmpty ||
-            coupon.excludedProductIds.isNotEmpty ||
-            coupon.excludedProductCategories.isNotEmpty {
-            return String.localizedStringWithFormat(Localization.shareMessageSomeProducts, amount, couponCode)
-        }
-        return String.localizedStringWithFormat(Localization.shareMessageAllProducts, amount, couponCode)
+        coupon.generateShareMessage(currencySettings: currencySettings)
```
Comment (itsmeichigo) [inline WooCommerce/Classes/ViewRelated/Coupons/CouponDetails/CouponDetailsViewModel.swift:100]: Great work for making this logic reusable! 👏

File: WooCommerce/Classes/ViewRelated/Coupons/CouponListViewController.swift [hunk 8]
```diff
@@ -179,6 +197,26 @@ private extension CouponListViewController {
         footerSpinnerView.stopAnimating()
         tableView?.tableFooterView = footerEmptyView
     }
+
+    /// Triggers the coupon creation flow
+    ///
+    func startCouponCreation(discountType: Coupon.DiscountType) {
+        let viewModel = viewModel.createAddEditCouponViewModel(with: discountType) { createdCoupon, shareMessage in
+            self.showCouponCreationSuccess(couponCode: createdCoupon.code, shareMessage: shareMessage)
+        }
```
Comment (itsmeichigo) [inline WooCommerce/Classes/ViewRelated/Coupons/CouponListViewController.swift]: We should be careful when capturing `self` strongly in closures. In this case, I think an `unowned` self is probably sufficient, but we can always fall back on using `weak` self for extra safety.

File: WooCommerce/Classes/ViewRelated/Coupons/CouponListViewController.swift [hunk 9]
```diff
@@ -179,6 +197,26 @@ private extension CouponListViewController {
...
+        present(addEditHostingController!, animated: true)
+    }
+
+    /// Display the Coupon creation success view
+    ///
+    func showCouponCreationSuccess(couponCode: String, shareMessage: String) {
+        let creationSuccessView = CouponCreationSuccess(couponCode: couponCode, shareMessage: shareMessage) {
+            self.dismiss(animated: true)
+            self.refreshCouponList()
+        }
```
Comment (itsmeichigo) [inline WooCommerce/Classes/ViewRelated/Coupons/CouponListViewController.swift]: Similar comment as above about capturing `self` strongly.

File: WooCommerce/Classes/ViewRelated/Coupons/CouponListViewController.swift (hunk 4 above)
Comment (itsmeichigo) [reply WooCommerce/Classes/ViewRelated/Coupons/CouponListViewController.swift]: Thanks for fixing this issue! However, I'm not sure if there can ever be the case when the local storage is not updated in time when the `onCouponDeleted` method is triggered. How about we update the view's state at the end of `buildCouponViewModels` - here we can be sure that the view models reflect the latest update of the local storage.

File: WooCommerce/Classes/ViewRelated/Coupons/CouponListViewController.swift [hunk 10]
```diff
@@ -319,10 +372,14 @@ private extension CouponListViewController {
...
+        let config: EmptyStateViewController.Config = .withButton(
             message: .init(string: Localization.emptyStateMessage),
-            image: .emptyCouponsImage
-        )
+            image: .emptyCouponsImage,
+            details: Localization.emptyStateDetails,
+            buttonTitle: Localization.createCouponAction
+        ) { button in
+            self.displayCouponTypeBottomSheet()
+        }
```
Comment (itsmeichigo) [inline WooCommerce/Classes/ViewRelated/Coupons/CouponListViewController.swift]: Let's capture `self` weakly here too.

File: WooCommerce/Classes/ViewRelated/Coupons/CouponListViewController.swift (hunk 8 above)
Comment (itsmeichigo) [reply WooCommerce/Classes/ViewRelated/Coupons/CouponListViewController.swift]: Also - I think we can simplify this part just so we don't have to keep a reference of `addEditHostingController` in this controller. Here's my suggestion:
- In `AddEditCouponViewModel`, when the creation is successful, we can ask the `AddEditCoupon` view to present the success view.
- When the success view is dismissed, it triggers a callback to let the `AddEditCoupon` know that it's time to get back to the list. The view will then trigger the view model's `onCompletion` closure, which asks the `CouponListViewController` to dismiss the hosting controller and reload the list. 

I see that this solution will probably cause 2 dismiss animations though 🤔 WDYT?

**Review by itsmeichigo — COMMENTED** [review]
Great work @ThomazFB! This works as expected 🎉  And also thanks for fixing the issue with the delete coupon feature!

I notice though that the coupon code field on the `AddEditCoupon` view is empty by default. If a user forgets to update this field, they'll get an error saying the creation fails due to empty coupon code. Should we prefill it with a random code initially to match the behavior on the web?


https://user-images.githubusercontent.com/5533851/173022491-9b28739c-1f0d-4dc9-9593-9288f02d0825.mp4

I also left a few comments regarding analytics, completion handling, along with some nitpicking below.

File: WooCommerce/Classes/ViewRelated/Coupons/CouponListViewController.swift [hunk 11]
```diff
@@ -216,7 +255,7 @@ private extension CouponListViewController {
     }
 
     func configureNavigationBarItems(hasCoupons: Bool) {
-        navigationItem.rightBarButtonItems = hasCoupons ? [searchBarButtonItem] : []
+        navigationItem.rightBarButtonItems = hasCoupons ? [createCouponButtonItem, searchBarButtonItem] : [createCouponButtonItem]
```
Comment (itsmeichigo) [inline WooCommerce/Classes/ViewRelated/Coupons/CouponListViewController.swift]: I forgot this part in the last review 😅 I think we should hide the coupon creation behind a feature flag, since this is still a work in progress. We also need to test this feature internally first so maybe it's safer not to show this to the public yet.

File: WooCommerce/Classes/ViewRelated/Coupons/CouponListViewController.swift (hunk 10 above)
Comment (itsmeichigo) [reply WooCommerce/Classes/ViewRelated/Coupons/CouponListViewController.swift]: We should also hide this behind a feature flag as well. How about duplicating `displayNoResultsOverlay` with another variant for when creation is enabled?

**Review by itsmeichigo — COMMENTED** [review]
I just realized that we forgot to create a feature flag for M3 😅 
Update: I added one in #7071

File: WooCommerce/Classes/ViewRelated/Coupons/CouponListViewController.swift (hunk 8 above)
Comment (ThomazFB) [reply WooCommerce/Classes/ViewRelated/Coupons/CouponListViewController.swift]: Following your suggestion, I actually managed to change this call stack in a way that doesn't trigger two dismiss animations, unless we use the swipe down gesture to dismiss the coupon creation success view since it will depend on the `onDisapper` SwiftUI event. Let me know what you think!

File: WooCommerce/Classes/ViewRelated/Coupons/Add and Edit Coupons/AddEditCouponViewModel.swift (hunk 3 above)
Comment (ThomazFB) [reply WooCommerce/Classes/ViewRelated/Coupons/Add and Edit Coupons/AddEditCouponViewModel.swift]: I totally missed the arguments detailed in your tracks P2 post, sorry 😅, it's fixed now!

**Comment by ThomazFB** [comment]
Hey @itsmeichigo! Thank you for such a detailed and insightful review! Took me a while to fix everything during the HACK week, but I've tackled everything you mentioned and the PR is now ready for a second round! 🙇

File: WooCommerce/Classes/ViewRelated/Coupons/Add and Edit Coupons/AddEditCoupon.swift [hunk 12]
```diff
@@ -92,217 +92,223 @@ struct AddEditCoupon: View {
...
                 ProductCategorySelector(isPresented: $showingSelectCategories,
-                                        viewConfig: categorySelectorConfig,
-                                        categoryListConfig: categoryListConfig,
-                                        viewModel: viewModel.categorySelectorViewModel)
+                        viewConfig: categorySelectorConfig,
+                        categoryListConfig: categoryListConfig,
+                        viewModel: viewModel.categorySelectorViewModel)
+            }
+            .sheet(isPresented: $viewModel.showingCouponCreationSuccess) {
+                let couponCode = viewModel.coupon?.code ?? ""
```
Comment (itsmeichigo) [inline WooCommerce/Classes/ViewRelated/Coupons/Add and Edit Coupons/AddEditCoupon.swift:308]: We should use a `guard let` here to make sure that the coupon code exists, otherwise log an error about the unexpected case. That will help a lot with catching the issue if it ever happens I believe.

File: WooCommerce/Classes/ViewRelated/Coupons/Add and Edit Coupons/AddEditCouponViewModel.swift [hunk 13]
```diff
@@ -233,7 +258,46 @@ final class AddEditCouponViewModel: ObservableObject {
...
+                ServiceLocator.analytics.track(.couponCreationSuccess)
+                self.coupon = coupon
+                self.onCompletion(result)
+                self.showingCouponCreationSuccess = true
+            case .failure(let error):
+                DDLogError("⛔️ Error creating the coupon: \(error)")
+                ServiceLocator.analytics.track(.couponCreationFailed, withError: error)
+                self.notice = NoticeFactory.createCouponErrorNotice(.other(error: error),
+                        editingOption: self.editingOption)
+            }
```
Comment (itsmeichigo) [inline WooCommerce/Classes/ViewRelated/Coupons/Add and Edit Coupons/AddEditCouponViewModel.swift]: It seems that we're triggering `onCompletion` only for the success case and not for failure. Maybe we could move the call to `onCompletion`  at the end of the switch to reuse it for both cases?

It also seems to me that there's the same error on `updateCoupon`, could you update that as well?

File: WooCommerce/Classes/ViewRelated/Coupons/Add and Edit Coupons/AddEditCouponViewModel.swift [hunk 14]
```diff
@@ -396,6 +460,16 @@ private extension AddEditCouponViewModel {
...
     }
 
+    func trackCouponCreateInitiated(with coupon: Coupon) {
+        ServiceLocator.analytics.track(.couponCreationInitiated, withProperties: [
+            "discount_type": coupon.discountType.rawValue,
+            "has_expiry_date": coupon.dateExpires != nil,
+            "includes_free_shipping": coupon.freeShipping,
+            "has_product_or_category_restrictions": coupon.excludedProductCategories.isNotEmpty || coupon.excludedProductIds.isNotEmpty,
+            "has_usage_restrictions": (coupon.usageLimit ?? 0) > 0
+        ])
```
Comment (itsmeichigo) [inline WooCommerce/Classes/ViewRelated/Coupons/Add and Edit Coupons/AddEditCouponViewModel.swift]: I think `has_description` is missing according to the tracking plan (p91TBi-7vL-p2). And also, `hasUsageRestrictions` should check for all the fields in the Usage Restrictions screen 😅

**Review by itsmeichigo — COMMENTED** [review]
Thanks @ThomazFB for the updates! This looks so much better now 👍 

One thing I notice is that if I turn on the switches for "Individual Use Only" and "Exclude Sale Items", these fields are still false after creating the coupon. The same problem happens to coupon editing as well, so I logged a separate issue for that: #7114.

There seem to be quite a few changes to the indents in `AddEditCoupon` file, I'm not sure why that happened. I also left other comments regarding analytics and other stuff. But overall the feature is looking good, so as soon as we get those issues out of the way, we can merge this 😄

File: WooCommerce/Classes/ViewRelated/Coupons/Add and Edit Coupons/AddEditCouponViewModel.swift (hunk 13 above)
Comment (ThomazFB) [reply WooCommerce/Classes/ViewRelated/Coupons/Add and Edit Coupons/AddEditCouponViewModel.swift]: Correct me if I'm wrong, but I don't see a clear reason for triggering the `onCompletion` closure for failure scenarios 🤔, if it's for dismissing as we do in the success case, we would be discarding the entire add / edit form merchant inputs for a possibly recoverable error, like sudden network unavailability or a server-side validation error that's worth a quick fix or retry, but besides dismissal, what else the calling view could take advantage of getting a `onCompletion` call for the failure scenario?

File: WooCommerce/Classes/ViewRelated/Coupons/Add and Edit Coupons/AddEditCoupon.swift (hunk 12 above)
Comment (ThomazFB) [reply WooCommerce/Classes/ViewRelated/Coupons/Add and Edit Coupons/AddEditCoupon.swift:308]: I'm not sure how we should add the `guard let` here 😅, the `sheet` declaration demands that at least a view is returned, so using the `guard let` triggers this build error here: 
image

Since the `showingCouponCreationSuccess` can only be true once a `.success` result is returned, I think that maybe we could move forward displaying the coupon creation success without displaying the coupon code and log the error instead of doing nothing after the creation response is sent to the `AddEditCouponViewModel`, WDYT? I think the creation success view deals nicely with this scenario:

Simulator Screen Shot - iPhone 13 Pro Max - 2022-06-21 at 12 07 06

Comment (itsmeichigo) [reply WooCommerce/Classes/ViewRelated/Coupons/Add and Edit Coupons/AddEditCoupon.swift:308]: You are right, I missed the fact that we are in a view builder here. Please proceed with your suggestion:

>  I think that maybe we could move forward displaying the coupon creation success without displaying the coupon code and log the error

Thanks!

File: WooCommerce/Classes/ViewRelated/Coupons/Add and Edit Coupons/AddEditCouponViewModel.swift (hunk 13 above)
Comment (itsmeichigo) [reply WooCommerce/Classes/ViewRelated/Coupons/Add and Edit Coupons/AddEditCouponViewModel.swift]: Looking at the implementation of the `onCompletion` closure used in the edit coupon case, I agree that we are using only the success case of this completion block. However, we are returning a `Result` in the closure, so the receiver of the closure will expect the failure case as well. 

I'd say if we proceed to handle the success case only, let's make the closure return only the created/updated coupon. We can maybe also rename the closure to `onSuccess` to better clarify the intention. WDYT?

File: WooCommerce/Classes/ViewRelated/Coupons/CouponListViewController.swift (hunk 8 above)
Comment (itsmeichigo) [reply WooCommerce/Classes/ViewRelated/Coupons/CouponListViewController.swift]: I missed this conversation yesterday 😅 Your solution looks good! 💯

File: WooCommerce/Classes/ViewRelated/Coupons/Add and Edit Coupons/AddEditCouponViewModel.swift (hunk 14 above)
Comment (ThomazFB) [reply WooCommerce/Classes/ViewRelated/Coupons/Add and Edit Coupons/AddEditCouponViewModel.swift]: Sorry for that 😰, it's fixed now here: https://github.com/woocommerce/woocommerce-ios/pull/7052/commits/fb7b16ed1dfd3f9e4658d1a7cb330f790f97a5c7

Let me know if it makes sense now!

File: WooCommerce/Classes/ViewRelated/Coupons/Add and Edit Coupons/AddEditCouponViewModel.swift (hunk 13 above)
Comment (ThomazFB) [reply WooCommerce/Classes/ViewRelated/Coupons/Add and Edit Coupons/AddEditCouponViewModel.swift]: Makes perfect sense! Fixed here: https://github.com/woocommerce/woocommerce-ios/pull/7052/commits/6859ac437f8e45f920e141c2cd34a168ac00db16

File: WooCommerce/Classes/ViewRelated/Coupons/Add and Edit Coupons/AddEditCoupon.swift (hunk 12 above)
Comment (ThomazFB) [reply WooCommerce/Classes/ViewRelated/Coupons/Add and Edit Coupons/AddEditCoupon.swift:308]: Nice! I've adjusted this part with the error log right here: https://github.com/woocommerce/woocommerce-ios/pull/7052/commits/0b265751b76e92bcc83cc281e41ce666f81d3ce5. But I had to do a hacky solution with the Log call using a wildcard declaration since the ViewBuilder doesn't allow me to declare a log function inside of that code 😅

**Comment by ThomazFB** [comment]
Hey @itsmeichigo, thanks for pointing out the identation issue around the `AddEditCoupon` and taking your time for a great review once again, I didn't notice it since the GitHub was hiding the large diff, this was my mistake because of my mixed usage of AppCode with Xcode, and I'm still sorting out how to make the same code format rules for both IDEs. The changes are now reverted. I've submitted changes for all your suggestions, and the PR is ready for another round!

File: WooCommerce/Classes/ViewRelated/Coupons/Add and Edit Coupons/AddEditCoupon.swift (hunk 12 above)
Comment (itsmeichigo) [reply WooCommerce/Classes/ViewRelated/Coupons/Add and Edit Coupons/AddEditCoupon.swift:308]: This definitely looks weird, but I cannot come up with a better solution so let's keep it as is 😅

File: WooCommerce/Classes/ViewRelated/Coupons/Add and Edit Coupons/AddEditCouponViewModel.swift [hunk 15]
```diff
@@ -396,6 +458,19 @@ private extension AddEditCouponViewModel {
...
+    func trackCouponCreateInitiated(with coupon: Coupon) {
+        ServiceLocator.analytics.track(.couponCreationInitiated, withProperties: [
+            "discount_type": coupon.discountType.rawValue,
+            "has_expiry_date": coupon.dateExpires != nil,
+            "includes_free_shipping": coupon.freeShipping,
+            "has_description": coupon.description.isNotEmpty,
+            "has_product_or_category_restrictions": coupon.excludedProductCategories.isNotEmpty || coupon.excludedProductIds.isNotEmpty,
+            "has_usage_restrictions": (coupon.usageLimit ?? 0) > 0
+            || (coupon.limitUsageToXItems ?? 0) > 0
+            || (coupon.usageLimitPerUser ?? 0) > 0
```
Comment (itsmeichigo) [inline WooCommerce/Classes/ViewRelated/Coupons/Add and Edit Coupons/AddEditCouponViewModel.swift]: For `has_usage_restrictions`, I'm thinking of something more complicated like what we did with [`checkUsageRestrictionsUpdated`](https://github.com/woocommerce/woocommerce-ios/blob/a99b74e251249f3a3e24a6e10ae4324f6b8cea56/WooCommerce/Classes/ViewRelated/Coupons/Add%20and%20Edit%20Coupons/AddEditCouponViewModel.swift#L332-L348) 😅 The idea is to cover all the restrictions defined there - does it make sense?

**Review by itsmeichigo — COMMENTED** [review]
@ThomazFB There's still one minor thing about tracking usage restrictions that I left in one comment below (please bear with me 😅)

File: WooCommerce/Classes/ViewRelated/Coupons/Add and Edit Coupons/AddEditCouponViewModel.swift (hunk 15 above)
Comment (ThomazFB) [reply WooCommerce/Classes/ViewRelated/Coupons/Add and Edit Coupons/AddEditCouponViewModel.swift]: Sorry for this sequence of mistakes 😅 , thanks for clarifying everything, it's fixed now here: https://github.com/woocommerce/woocommerce-ios/pull/7052/commits/69d7dc809915e50808c685a46795703c6be1b178. Let me know if that's what you had in mind.

**Review by itsmeichigo — APPROVED** [review]
Thank you @ThomazFB for your patience! This looks good and works as expected now 🎉 🚀
